hitman3.asl

```diff
@@ -1,58 +1,356 @@
+    }.ToList();
+
+    //Version check if needed in future to change signatures
+    //
+    /*	
+	var versionInfo = FileVersionInfo.GetVersionInfo(mainModule.FileName);
+	if(versionInfo.FileVersion == "3.110.1.0")
+	{
+	
+	
+	}
+	*/
+
+    var e_lfanew = memory.ReadValue<int>(baseAddress + 0x3C);
+    var SizeOfOptionalHeader = memory.ReadValue<short>(baseAddress + e_lfanew + 0x4 + 0x10);
+    var textSectionBase = baseAddress + e_lfanew + 0x18 + SizeOfOptionalHeader;
+
+    var textVirtualSize = memory.ReadValue<int>(textSectionBase + 0x08);
+    var textVirtualAddress = memory.ReadValue<int>(textSectionBase + 0x0C);
+
+    var textSectionData = memory.ReadBytes(baseAddress + textVirtualAddress, textVirtualSize);
+
+    //We probably could use internaly class SignatureScanner but it dosen't support multiple signatures at the same time
+    //
+
+    for (var i = 0; i < textSectionData.Length; i++)
+    {
+        bool finish = true;
+
+        foreach (var signature in signaturesList)
+        {
+            if (signature.state[0] == 1)
+                continue;
+
+            finish = false;
+
+            var currentIndex = signature.state[1];
+
+            if (signature.sig[currentIndex] == 0xFF
+            || signature.sig[currentIndex] == textSectionData[i])
+            {
+                if ((currentIndex + 1) == signature.sig.Count)
+                {
+                    var sigAddress = baseAddress + textVirtualAddress + (i - currentIndex);
+
+                    if (signature.type == "offset")
+                    {
+                        signature.state[2] = memory.ReadValue<int>(sigAddress + signature.read);
+                    }
+                    else if (signature.type == "pointer")
+                    {
+                        signature.state[2] = memory.ReadValue<int>(sigAddress + signature.read) + textVirtualAddress + (i - currentIndex) + 0x7;
+                    }
+
+                    signature.state[0] = 1;
+                }
+                signature.state[1]++;
+                continue;
+            }
+
+            signature.state[1] = 0;
+        }
+
+        if (finish)
+            break;
+    }
+
+    var gameAddresses = new Dictionary<string, int>();
+
+    foreach (var signature in signaturesList)
+    {
+        gameAddresses[signature.name] = signature.state[2];
+    }
+
+    vars.timePointer = new DeepPointer(gameAddresses["gameTime"]);
+    vars.hudMissionTimeControllerPointer = new DeepPointer(gameAddresses["hudMissionTimeController"]);
+    vars.missionTimePointer = new DeepPointer(gameAddresses["hudMissionTimeController"], 0x40);
+    vars.gateExitedPointer = new DeepPointer(gameAddresses["contractsManager"] + gameAddresses["gateExited"]);
+    vars.metadataLocationPointer = new DeepPointer(gameAddresses["contractsManager"] + gameAddresses["metadataLocation"], 0x00);
+
+    //I don't know how to modify pointers in state so will create my own variables
```

Comment:
don't think you can. Afaik states are static, usually you'd have multiple versions with static pointers/offsets and just detect the relevant one rather than scanning signatures on init each time. But the scan likely isn't gonna take that long going through just the game address space so I think it's fine... ofc it does make us reimplement our own old/new state tracking though. Maybe it would be better to just write a script that does the scan for the variables and dumps out a relevant "state" object code that could be easily added on each game update? Idk. Both have tradeoffs.

Reply:
> don't think you can. Afaik states are static, usually you'd have multiple versions with static pointers/offsets and just detect the relevant one rather than scanning signatures on init each time. But the scan likely isn't gonna take that long going through just the game address space so I think it's fine... ofc it does make us reimplement our own old/new state tracking though. Maybe it would be better to just write a script that does the scan for the variables and dumps out a relevant "state" object code that could be easily added on each game update? Idk. Both have tradeoffs.

I agree with the sentiment: the signature approach is cool to survive patches (assuming sigs don't change, which they do seem stable so far). On the other hand it adds a lot of complexity to handle the signatures (see my other comment) and a bit of complexity to deal with these custom state variables.

Reply:
I like the signature approach but yeah you are right, that custom state variables makes me pain, that is why i added comment "Please fix me" because probably there is a better way for that

Reply:
Regarding the custom vars, I found two alternatives:

* per AutoSplitters [README](https://github.com/LiveSplit/LiveSplit.AutoSplitters/blob/master/README.md#vars) you can also set variables on the `current` object, which takes care of moving the last value to `old` in the update method automatically. I tried it by replacing `vars.currentTime` with `current.time`. Then you can use `old.time` instead of `vars.oldTime` and there's no need for the `vars.oldTime = vars.currentTime` in `update`.
* Perhaps even better, there's [MemoryWatcher](https://github.com/LiveSplit/LiveSplit/blob/master/LiveSplit/LiveSplit.Core/ComponentUtil/MemoryWatcher.cs). You can initialize it with DeepPointers and call its Update methods in `update`. Then it keeps track of `current` and `old` value at that memory location. Example usage: https://raw.githubusercontent.com/jbzdarkid/Autosplitters/master/LiveSplit.TheWitness2.asl

By the way, I'm also fine if we do such improvements in a second PR and focus on getting the current version fully working first (e.g. Steam).

Reply:
Steam version fixed, and I changed old/current system to MemoryWatcher